### ISSUE TITLE
Feature request: visualizing real algebraic curves and surfaces with surf and surfer

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#1859]
[surf](http://surf.sourceforge.net/) is a program for visualizing real algebraic curves and surfaces, with an optional Gtk gui:
![image](https://user-images.githubusercontent.com/3147718/105524170-e4537a00-5ca4-11eb-830e-292e95a974c8.png)

[SURFER](https://imaginary.org/program/surfer) is a Java applet for similar purposes, but is newer and has other features:
![image](https://user-images.githubusercontent.com/3147718/105524426-3eecd600-5ca5-11eb-97f6-722f0deb15d5.png)

Would be good to support visualizing objects over `RR` this way. See https://www.singular.uni-kl.de/Manual/4-0-3/sing_1812.htm for an example.

**Comment by mahrud** [comment]
Added images.

**Comment by pzinn** [comment]
can it run in the browser?

**Comment by mahrud** [comment]
The latter is a Java applet, so maybe. You could also make it work with
Cinderella and CindyJS, instead.

**Comment by mahrud** [comment]
Here's an example of CindyJS: https://math.umn.edu/~mahrud/experiments/surfaces/

**Comment by mikestillman** [comment]
That looks beautiful!

One thing I really liked about an older version of surf, was the quality of implicit plane curves it could create, and how beautiful it did that.  Alas, I think that code was not retained in later versions of surf (in 2001, I had code for interfacing the old version of surf with Macaulay2, but this was before packages, and that code went to seed, although I probably still have it).  Does surfer/CindyJS handle implicit curves F(x,y) = 0?

**Comment by 8d1h** [comment]
Hi! I was playing around with the M2 jupyter kernel, and I found out that it's fairly easy to display *static* images in a jupyter notebook using surf. [Here](https://github.com/8d1h/Macaulay2-Jupyter-Kernel/blob/master/demo/surf-plot.ipynb) is a demo notebook. All the M2 code required is in the first cell (and in the command line one can use `show` instead to view the image). Should be easy to add to the web version provided that the surf binary is available on the server.

The CindyJS version is really impressive though!

**Comment by mahrud** [comment]
Oh that looks excellent! Yes, I think we should implement `(show, Ideal)` which either creates and displays an image, or possibly opens a tab using CindyJS. Your function looks like a great place to start a new package.

**Comment by pzinn** [comment]
quickly tried your code. for the fedora packaged version of surf I had to add the option "--no-gui".
for the rest works out of the box. note that the package Text includes an IMG type which might simplify a bit the html stuff.
![surf](https://user-images.githubusercontent.com/4632711/110753203-679f4e00-829a-11eb-97d3-7c9dbe18be39.png)

**Comment by pzinn** [comment]
it might be more clever to have `surf` call a script which adds this "-no-gui" automatically, just realized.
to myself, so I don't forget what I did:
```
#!/bin/bash
/usr/lib64/surf-geometry/surf --no-gui $@
```

**Comment by 8d1h** [comment]
That's great! Yes you are right, I compiled surf using `--disable-gui` so I don't need to use the option `--no-gui`.

BTW, if we want to make this into a package, it might be a good idea to separate the first part into a universal interface for displaying images, which will work well both for the web version and locally. Surely there are other packages that need to show images too.

**Comment by d-torrance** [comment]
> Surely there are other packages that need to show images too.

`Graphs` and `Posets` both do, off the top of my head.

**Comment by pzinn** [comment]
yes, the precompiled version for fedora may not be so great since it has lots of dependencies which are only used by the gui.

> That's great! Yes you are right, I compiled surf using `--disable-gui` so I don't need to use the option `--no-gui`.
> 
> BTW, if we want to make this into a package, it might be a good idea to separate the first part into a universal interface for displaying images, which will work well both for the web version and locally. Surely there are other packages that need to show images too.

**Comment by pzinn** [comment]
so I tried to compile surf with `--disable-gui` today, no luck. I'm getting
```
make[1]: *** No rule to make target 'gtkgui/start.o', needed by 'surf'.  Stop.
```
which I'm assuming shouldn't happen with `--no-gui`.

**Comment by pzinn** [comment]
I guess the Makefile was somehow corrupted. fixed the problem manually.
expect surf to be available some time next week on Macaulay2Web :)

**Comment by 8d1h** [comment]
> so I tried to compile surf with `--disable-gui` today, no luck. I'm getting
> 
> ```
> make[1]: *** No rule to make target 'gtkgui/start.o', needed by 'surf'.  Stop.
> ```
> 
> which I'm assuming shouldn't happen with `--disable-gui`.

Yes I realized that I originally used a [patched version](http://www-ftp.lip6.fr/pub/math/sagemath/spkg/upstream/surf/index.html) hosted on the server of Sage, which removed the gui entirely. To use the official version on sourceforge one has to fix it manually (and also another problem with the min/max macro). If anyone uses Arch Linux, I packaged it on [AUR](https://aur.archlinux.org/packages/surf-geometry/).

> expect surf to be available some time next week on Macaulay2Web :)

That's great!

**Comment by mahrud** [comment]
Does the patched version also accept `--no-gui`? Would be great to be able to use the packaged version for fedora, ubuntu, etc.

**Comment by d-torrance** [comment]
On Debian-based distros, surf is compiled with `--disable-gui` and available as `surf-alggeo`.

I just tested 8d1h's code and it works great with the Ubuntu 20.10 package with a couple tweaks (`surf` -> `surf-alggeo` and appending `.jpg` to the `filename` specified in `INPUT`).

**Comment by 8d1h** [comment]
> Does the patched version also accept `--no-gui`? Would be great to be able to use the packaged version for fedora, ubuntu, etc.

The patched version from Sage removed gui related stuffs from Makefile entirely and thus can only be compiled using `--disable-gui`. In this case passing `--no-gui` or not has no effect at all, so yes I think we can always pass the `--no-gui` option.

Fedora distributes surf as [`surf-geometry`](https://src.fedoraproject.org/rpms/surf-geometry), as @pzinn already mentioned.

> I just tested 8d1h's code and it works great with the Ubuntu 20.10 package with a couple tweaks (`surf` -> `surf-alggeo` and appending `.jpg` to the `filename` specified in `INPUT`).

That's weird. Actually I had some issues when plotting with `surf` in `Singular`, since it appends `.jpg` to the file name and I end up getting an output `surf.jpg.jpg`. Does this somehow depend on the distro?

**Comment by d-torrance** [comment]
> > Does the patched version also accept `--no-gui`? Would be great to be able to use the packaged version for fedora, ubuntu, etc.
> 
> The patched version from Sage removed gui related stuffs from Makefile entirely and thus can only be compiled using `--disable-gui`. In this case passing `--no-gui` or not has no effect at all, so yes I think we can always pass the `--no-gui` option.
> 
> Fedora distributes surf as [`surf-geometry`](https://src.fedoraproject.org/rpms/surf-geometry), as @pzinn already mentioned.
> 
> > I just tested 8d1h's code and it works great with the Ubuntu 20.10 package with a couple tweaks (`surf` -> `surf-alggeo` and appending `.jpg` to the `filename` specified in `INPUT`).
> 
> That's weird. Actually I had some issues when plotting with `surf` in `Singular`, since it appends `.jpg` to the file name and I end up getting an output `surf.jpg.jpg`. Does this somehow depend on the distro?

The Debian package adds a [patch](https://salsa.debian.org/science-team/surf-alggeo/-/blob/master/debian/patches/upstream-automatic_extension.patch) that deals with this.  By default, extensions are not added automatically, but if `--enable-auto-extension` is used, then they are.

**Comment by 8d1h** [comment]
> The Debian package adds a [patch](https://salsa.debian.org/science-team/surf-alggeo/-/blob/master/debian/patches/upstream-automatic_extension.patch) that deals with this. By default, extensions are not added automatically, but if `--enable-auto-extension` is used, then they are.

Thanks, that explains it.

**Comment by pzinn** [comment]
surf is now available at https://www.unimelb-macaulay2.cloud.edu.au/.

**Comment by mahrud** [comment]
Would be good if `view` was added as a synonym for `show`, and they visualized plane and space curves.
